README.md
![Image description](Documentation~/images/BarracudaLanding.png)

# Unity Barracuda

**Unity Barracuda** is a lightweight **cross-platform** Neural Networks **inference library for Unity**. \
Barracuda can run Neural Networks both on GPU and CPU. For details, please look for [Supported Platforms](Documentation~/SupportedPlatforms.md).

**Note:** The Barracuda package has been replaced by the `Sentis` package, which is in a closed beta phase. Refer to the [Sentis documentation](https://docs.unity3d.com/Packages/com.unity.sentis@1.0) for more information.
You can [sign up for the closed beta](https://create.unity.com/ai-beta).

Currently Barracuda is production-ready for use with [machine learning (ML) agents](https://github.com/Unity-Technologies/ml-agents) and number of [other network architectures](Documentation~/SupportedArchitectures.md). When you use Barracuda in other scenarios, it is in the preview development stage.


# Installation

[Installing Barracuda](Documentation~/Installing.md) goes through how to install Barracuda, both locally and remotely. 

# Reporting issues

If you encounter issues running Barracuda in your Unity project, please report them on our [GitHub repo](https://github.com/Unity-Technologies/barracuda-release/issues).


CHANGELOG.md
# Changelog
All notable changes to this package will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [3.0.01] - 2023-06-28
### Changed
- Update documentation

## [3.0.0] - 2022-01-24
### Changed
- Combined verified release of 2.1.x-2.4.x improvements

## [2.4.0] - 2021-12-14
### Added
- Added `ModelLoader.LoadAsync` methods for coroutine-based model loading
- Added support for Model weights as half on disk and in memory for CPU backend.
- Vastly improved operator coverage for PixelShader worker.
- Added YoloV3Tiny demo to [Starter Kit](https://github.com/Unity-Technologies/barracuda-starter-kit)

### Changed
- Added `flipY` as an optional parameter to `TensorToRenderTexture`
- Download from texture use pixel shader when compute shader are not available.
- Improved support for dynamic input shape and rank inference.

### Fixed
- Fixed `Resample2D` when using input tensor as size parameter.
- Fixed `OneHot` when using rank 2 input tensor.
- Fix TextureToTensor for PixelShader backend.
- Fix Added importer support for ConvTranspose with 1D spatial inputs.

## [2.3.1] - 2021-10-26
### Fixed
- Fixed `CSharpBurst` backend failure on `RandomNormal` and `RandomUniform` ops

## [2.3.0] - 2021-10-15
### Added
- PixelShader worker! Allowing to run on GPU without requiring support for compute shaders. Note: limited to 4D tensors for now.
- ONNX: Added support for `transA` and `transB` parameters for `Gemm` operator.
- ONNX: Added support for multi-axis `Squeeze`/`Unsqueeze`.
- Added support for `RoiAlign`.

### Changed
- Improved Tensor storage handling for Burst backend, should reduce overall memory pressure when used.
- Improved `Transpose` performance when they are not changing memory layout.
- Improved automatic test coverage for consoles.
- Improved `ConvTranspose2D` performance for large spatial kernel.

### Fixed
- Added ability to create tensor from RenderTexture when compute shaders or GPU are not available.

## [2.2.1] - 2021-09-14
### Fixed
- Fixed `CSharpBurst` Random layer failure when model is loaded from an old .nn file

## [2.2.0] - 2021-09-01
### Added
- Added Burst implementation for Random layers, improves scheduling of the whole network computation
- Added scale and bias options for `TensorFromTexture`
- Added `Erf` layer support
- Added `LogSoftmax` axis support

### Changed
- Improved MatMul performance on `CSharpBurst` backend on Desktop platforms
- Improved temp Tensor allocation handling, should reduce GC pressure
- Improved Tensor storage handling for PrecompiledCompute backend, should reduce overall memory pressure when used.
- Exposed flag for GPU worker to take ownership of weights, reduces memory pressure on CPU side, but limits `Model` object re-use.
- Improved convolution performance on Mobiles
- Improved 1x1 convolution performance
- Minimal supported Unity version updated to 2019.4.29f1
- Improved Barracuda installation instructions 

### Fixed
- Fixed broadcasting between rank 4 and rank 2 operands 
- Fixed GPU `Flatten` when channel order is NHWC

## [2.1.0] - 2021-05-17
### Added
- ONNX: Added support for the `metadata_props` field, which is exposed as a <string,string> dictionary via `Model.Metadata`

### Changed
- Improved `Slice` range handling
- Empty tensor buffers are now filled with `-1` instead of the random data, helps to reduce accidental errors
- Greatly reduced memory allocation when using GPU backends (no more CPU mem cache for intermediate tensors). Mostly helps mobiles with unified memory architecture
- Improved `SoftMax` performance
- Improved `LSTM` performance on CPU

### Fixed
- Fixed default axis to `channels` in `ModelBuilder.SoftMax`
- Fixed `Optimize Model` checkbox in model importer UI
- Fixed crash when tensor is created on CPU and later is moved to GPU
- Fixed out of bounds write in NCHW path
- Fixed crash in Tensor finalizers when using `CSharpBurst` backend
- Fixed Gather behavior for rank3 tensors

### Known issues
- There is small regression in ARMv7 performance when running `CSharpBurst` backend. It can be worked around on most of the phones by shipping ARMv8(ARM64) builds

## [2.0.0] - 2021-04-05
### Changed
- Combined verified release of 1.1.x-1.4.x improvements
- Burst updated to 1.6.0
- Now Barracuda requires Unity 2019.4.29f1 or newer 

### Fixed
- Fixed Transpose removal pass for ML-Agents networks

## [1.4.0] - 2021-04-01
### Added

Tools/onnx_to_barracuda.py
from __future__ import print_function
import numpy as np
import onnx
import struct
import barracuda
from barracuda import Struct
from google.protobuf.json_format import MessageToJson
from collections import Iterable
from pprint import pprint

if __name__ == '__main__':
    # Handle command line argumengts
    args = barracuda.parse_args(
        description = 'Convert ONNX model to Barracuda binary',
        source_extension = '.onnx',
        help = 'input ONNX serialized  file')
    # The following code can be used as an example of API used from another module
    # convert() is the main entry point for converter
    import onnx_to_barracuda as onnx2bc
    onnx2bc.convert(args.source_file, args.target_file, args.trim_unused_by_output, args)


# ONNX format:  # See: https://github.com/onnx/onnx/blob/master/onnx/onnx.proto
# ONNX schema: https://github.com/onnx/onnx/blob/master/docs/Operators.md
# ONNX conventions: https://github.com/onnx/onnx/blob/master/docs/OpConventions.md

def get_epsilon(layer):
    # default epsilon taken from https://github.com/onnx/onnx/blob/master/docs/Operators.md#BatchNormalization
    return get_attr(layer, 'epsilon', default=1e-05)

known_classes = {
    'Gemm': Struct(id = 1,
                    patch_shapes = lambda shapes: [
                        shape_to_HW(shapes[0]),                 # W
                        bias(shape_to_HW(shapes[-1])),          # B
                    ],
                    patch_data = lambda data: [
                        data_to_HW(data[0]),
                        data[1]
                    ]),
    'MatMul': Struct(id = 1,
                    patch_shapes = lambda shapes: [
                        shape_to_HW(shapes[0]),                 # W
                        bias(shape_to_HW(shapes[-1])),          # ZERO
                    ],
                    patch_data = lambda data: [
                        data_to_HW(data[0]),
                        np.zeros(np.shape(data[1]))
                    ]),

    'Conv': Struct(
                    id = lambda layer:
                        21 if is_depthwise_convolution(layer) else # DepthwiseConv2D
                        20, # Conv2D
                    patch_shapes = lambda shapes: [
                        shape_to_HWCK(shapes[0]),               # K 
                        bias(shape_to_HWCK(shapes[-1]))         # B
                    ],
                    patch_data = lambda data: [
                        data_to_HWCK(data[0]),
                        data[1]
                    ]), # TODO: args(group!=0|kernel_shape.channels)
    'ConvTranspose': Struct(
                    id = 22,
                    patch_shapes = lambda shapes: [
                        shape_to_HWKC(shapes[0]),               # K 
                        bias(shape_to_HWKC(shapes[-1]))         # B
                    ],
                    patch_data = lambda data: [
                        data_to_HWKC(data[0]),
                        data[1]
                    ]), # TODO: args(output_padding, output_shape)
    'Upsample':         23, 'ResizeNearest':    23,

    'MaxPool':          25,
    'AveragePool':      26, # TODO: args(count_include_pad)
    'GlobalMaxPool':    27,
    'GlobalAveragePool':28,
    'Pad':              29, # TODO: args(mode), implemented as Border2D
    'Crop': Struct(id = 29, # implemented as Border2D
                    patch_attrs = lambda attrs, layer: {
                        'pads': border_to_inverse_padding(layer)
                    }),
    # TODO: 3D

    'Activation':       50,
    'BatchNormalization': Struct(
        # TODO: args(epsilon)
                    id = 51, # after fusion implemented as ScaleBias
                    patch_shapes = lambda shapes: [
                        [1, 1, 1, shapes[0][0]],             # S
                        [1, 1, 1, shapes[0][0]],             # B
                    ],
                    patch_data = lambda data, layer:
                        # fuse [gamma, beta, mean, var] => [scale, bias]
                        barracuda.fuse_batchnorm_weights(data[0], data[1], data[2], data[3], get_epsilon(layer)) if len(data) == 4 else
                        # fuse [ONE, beta, mean, var] => [scale, bias]
                        barracuda.fuse_batchnorm_weights(np.ones(np.shape(data[0])), data[0], data[1], data[2], get_epsilon(layer))
                    ),
    'BatchNormalizationRuntime': Struct(
                    id = 52,
                    patch_shapes = lambda shapes: [
                        [1, 1, 1, shapes[0][0]],             # G
                        [1, 1, 1, shapes[0][0]],             # B
                    ],
                    patch_data = lambda data:
                        [data[0], data[1]] if len(data) == 4 else
                        [np.ones(np.shape(data[0])), data[0]]
                    ),
    'InstanceNormalization': Struct(
                    id = 52,
                    patch_shapes = lambda shapes: [
                        [1, 1, 1, shapes[0][0]],             # G
                        [1, 1, 1, shapes[0][0]],             # B
                    ],
                    patch_data = lambda data:
                        [data[0], data[1]] if len(data) == 2 else
                        [np.ones(np.shape(data[0])), data[0]]
                    ),
    'LRN':              53, # TODO: args(beta, bias)

    'DropoutRuntime':   60,
    'RandomNormalLike': 64,
    'RandomUniformLike':65,
    'Multinomial':      66,

    'Add':              100,
    'Sum':              100,

    'Sub':              101,
    'Mul':              102,
    'Div':              103,
    'Pow':              104,

    'Min':              110,
    'Max':              111,

    'ReduceL1':         120, # TODO: args(axes)
    'ReduceL2':         121, # TODO: args(axes)
    'ReduceLogSum':     122, # TODO: args(axes)
    'ReduceLogSumExp':  123, # TODO: args(axes)
    'ReduceMax':        124, # TODO: args(axes)
    'ReduceMean':       125, # TODO: args(axes)
    'ReduceMin':        126, # TODO: args(axes)
    'ReduceProd':       127, # TODO: args(axes)
    'ReduceSum':        128, # TODO: args(axes)
    'ReduceSumSquare':  129, # TODO: args(axes)

    'Greater':          140, # TODO: args(axes)
    'GreaterEqual':     141, # TODO: args(axes)
    'Less':             142, # TODO: args(axes)
    'LessEqual':        143, # TODO: args(axes)
    'Equal':            144, # TODO: args(axes)
    'Or':               145, # TODO: args(axes)
    'And':              146, # TODO: args(axes)
    'Not':              147, # TODO: args(axes)
    'Xor':              148, # TODO: args(axes)

    'Flatten':          200,
    'Reshape':          201,
    'Transpose':        202,
    'Squeeze':          203, # TODO: args(axes)
    'Unsqueeze':        204, # TODO: args(axes)

    'Concat':           210,
    'Slice':            211, # TODO: args(axes, ends, starts)
    'Tile':             212  # TODO: args(repeats)
}

requires_runtime_flag = {
    'Dropout' : 'DropoutRuntime',
    'BatchNormalization' : 'BatchNormalizationRuntime',
}

known_activations = {
    'Linear' : 0,       # dummy, doesn't exist in ONNX standard, all nodes have no activation by default
    'Identity' : 0,
    'Relu' : 1,
    'Softmax' : 2,      # TODO: args(axis)
    'Tanh' : 3,
    'Sigmoid' : 4,
    'Elu' : 5,
    'Relu6' : 6,        # doesn't exist in ONNX standard yet
    'LeakyRelu' : 7,
    'Selu' : 8,         # TODO: args(gamma)
    'Swish' : 9,        # doesn't exist in ONNX standard yet

    'LogSoftmax': 10,   # TODO: args(axis)
    'Softplus': 11,
    'Softsign': 12,

    'Hardmax': 20,      # TODO: args(axis)
    'HardSigmoid': 21,

    'Abs': 100,
    'Neg': 101,
    'Ceil': 102,
    'Clip': 103,
    'Floor': 104,

    'Reciprocal': 110,
    'Sqrt': 111,
    'Pow': 112,
    'Exp': 113,
    'Log': 114,

    'Acos': 200,
    'Acosh': 201,
    'Asin': 202,
    'Asinh': 203,
    'Atan': 204,
    'Atanh': 205,
    'Cos': 206,
    'Cosh': 207,
    'Sin': 208,
    'Sinh': 209,
    'Tan': 210
}

known_paddings = {
    'VALID' : [0,0,0,0],
    'SAME_UPPER'  : [-1],
    'SAME_LOWER'  : [-2],
}

# Helper
def embody(v, default=0):
    return default if v is None else v

# Parse
def get_attr(node, attr_name, default=None):
    for attr in node.attribute:
        if attr.name == attr_name:
            if attr.AttributeType.Name(attr.type) == 'INTS':
                return attr.ints
            elif attr.AttributeType.Name(attr.type) == 'FLOATS':
                return attr.floats
            elif attr.AttributeType.Name(attr.type) == 'STRINGS':
                return attr.strings
            elif attr.AttributeType.Name(attr.type) == 'INT':
                return attr.i
            elif attr.AttributeType.Name(attr.type) == 'FLOAT':
                return attr.f
            elif attr.AttributeType.Name(attr.type) == 'STRING':
                return attr.s.decode("utf-8")
            elif attr.AttributeType.Name(attr.type) == 'TENSOR':
                return attr.t
    return default

def is_depthwise_convolution(layer):
    # https://github.com/onnx/onnx/blob/master/docs/Operators.md#conv
    # depthwise convolutions have multiple channel groups (in_channels == group_count)
    return get_attr(layer, 'group', default=1) > 1

def border_to_inverse_padding(layer):
    border = get_attr(layer, 'border', np.ones(4)) # 'border' is (leftBorder, topBorder, rightBorder, bottomBorder).
    height, width = get_attr(layer, 'scale', np.ones(2)) # 'scale' is (height, width)
    return [-border[0] * width, -border[1] * height, -border[2] * width, -border[3] * height]

def axis_to_NHWC(axis):
    """ Convert from NCHW => NHWC
    """
    assert(axis >= -1 and axis < 4)
    return [-1, 0, 3, 1, 2][axis+1];

def shape_to_NHWC(shape):
    """ Convert from NCHW|NCW => NHWC
    """
    if len(shape) == 4: # NCHW => NHWC
        return [shape[0], shape[2], shape[3], shape[1]]
    if len(shape) == 3: # NCW => N1WC
        return [shape[0], 1, shape[2], shape[1]]
    return shape # Not NCHW, return as is

def shape_to_HW(shape):
    """ Convert from WH => HW
    """
    if len(shape) != 2:
        return shape # Not WH, return as is
    return [shape[1], 1, 1, shape[0]]

def shape_to_HWCK(shape):
    """ Convert from KCHW|KCW => HWCK
    """
    if len(shape) == 4: # KCHW => HWCK
        return [shape[2], shape[3], shape[1], shape[0]]
    if len(shape) == 3: # KCW => 1WCK
        return [1, shape[2], shape[1], shape[0]]
    return shape # Not KCHW, return as is

def shape_to_HWKC(shape):
    """ Convert from KCHW|KCW => HWKC
    """
    if len(shape) == 4: # KCHW => HWKC
        return [shape[2], shape[3], shape[0], shape[1]]
    if len(shape) == 3: # KCW => 1WKC
        return [1, shape[2], shape[0], shape[1]]
    return shape # Not KCHW, return as is

def bias(shape):
    return [1, 1, 1, shape[-1]]

def adapt_input_shape(shape):
    if hasattr(shape, 'dim'):
        shape = [dim.dim_value for dim in shape.dim]
    if len(shape) == 1:
        return bias(shape)
    if len(shape) == 2:
        return shape_to_HW(shape)
    if len(shape) == 3 or len(shape) == 4:
        return shape_to_NHWC(shape)
    return shape

def data_to_HW(tensor):
    """ Convert from WH => HW
    """
    return np.transpose(tensor)

def data_to_HWCK(tensor):
    """ Convert from KCHW|KCW => HWCK
    """
    if len(tensor.shape) == 3:
        tensor = np.expand_dims(tensor, axis=2)
    return np.transpose(tensor, (2,3,1,0))
    
def data_to_HWKC(tensor):
    """ Convert from KCHW|KCW => HWKC
    """
    if len(tensor.shape) == 3:
        tensor = np.expand_dims(tensor, axis=2)
    return np.transpose(tensor, (2,3,0,1))

def get_tensor_data(tensor):
    elems = np.product(tensor.dims)
    if tensor.raw_data:
        floats = struct.unpack('<'+str(int(elems))+'f', tensor.raw_data)
    elif tensor.float_data:
        floats = tensor.float_data
    elif tensor.int32_data:
        floats = np.array(tensor.int32_data, dtype=float)
    elif tensor.int64_data:
        floats = np.array(tensor.int64_data, dtype=float)
    return np.array(floats).reshape(tensor.dims)

def flatten(items,enter=lambda x:isinstance(x, list)):
    # http://stackoverflow.com/a/40857703
    # https://github.com/ctmakro/canton/blob/master/canton/misc.py
    """Yield items from any nested iterable; see REF."""
    for x in items:
        if enter(x):
            yield from flatten(x)
        else:
            yield x

def replace_strings_in_list(array_of_strigs, replace_with_strings):
    "A value in replace_with_strings can be either single string or list of strings"
    potentially_nested_list = [replace_with_strings.get(s) or s for s in array_of_strigs]
    return list(flatten(potentially_nested_list))

#########################################################

class ModelBuilderContext:
    def __init__(self):
        self.layers = []
        self.input_shapes = {}
        self.model_tensors = {}
        self.model_memories = []
        self.map_ignored_layer_to_its_input = {}

def process_model(model, args):
    o_context = ModelBuilderContext()

    o_context.input_shapes = { i.name:adapt_input_shape(i.type.tensor_type.shape) for i in model.graph.input }
    o_context.model_tensors = { tensor.name:tensor for tensor in model.graph.initializer }

    for layer in model.graph.node:
        process_layer(layer, o_context, args)
   
    return o_context.layers, o_context.input_shapes, o_context.model_tensors, o_context.model_memories

def process_layer(layer, context, args):
    model_tensors = context.model_tensors
    map_ignored_layer_to_its_input = context.map_ignored_layer_to_its_input

    name = layer.output[0] if len(layer.output) > 0 else layer.name # prefer node.output over the node.name
    class_name = layer.op_type
    inputs = layer.input # ONNX inputs are always explicit, but in case of Keras we had 'inputs = layer.input or [prev_layer_name]'
    inputs = replace_strings_in_list(inputs, map_ignored_layer_to_its_input)

    if class_name == 'Constant':
        model_tensors[name] = get_attr(layer, 'value')
        model_tensors[name].name = name
        #print('CONST:', name, model_tensors[name].dims, struct.unpack('<'+str(np.prod(model_tensors[name].dims))+'f', model_tensors[name].raw_data))
        return

    if args.print_layers or args.verbose:
        print("'%s' %s %s" % (name, class_name, inputs))

    if class_name in known_activations:
        activation = class_name
        class_name = 'Activation'
    else:
        activation = 'Linear'
   
    if not class_name in known_classes:
        if class_name in requires_runtime_flag:
            print('SKIP:', class_name, 'layer is used only for training')
        else:
            print('IGNORED:', class_name, 'unknown layer')
        map_ignored_layer_to_its_input[name] = inputs
        return

    klass = known_classes[class_name]
    if type(klass) == int:
        klass = Struct(id = klass)

    o_l = Struct()
    o_l.type = klass.id(layer) if callable(klass.id) else klass.id
    o_l.class_name = class_name
    o_l.name = name

    axis        = axis_to_NHWC(get_attr(layer, 'axis', -1))
    auto_pad    = get_attr(layer, 'auto_pad')
    pads        = get_attr(layer, 'pads')
    strides     = get_attr(layer, 'strides')
    pool_size   = get_attr(layer, 'kernel_shape')
    shape       = get_attr(layer, 'shape')
    starts      = get_attr(layer, 'starts')
    ends        = get_attr(layer, 'ends')
    slice_strides = [1,1,1,1] if starts and ends else []
    #TODO properly extract scale from const Tensor for Upsample layers
    size        = [get_attr(layer, 'height_scale'), get_attr(layer, 'width_scale')] if get_attr(layer, 'width_scale') and class_name == 'Upsample' else [2,2]
    alpha       = get_attr(layer, 'alpha') or get_attr(layer, 'ratio') or get_attr(layer, 'value')
    beta        = get_attr(layer, 'beta') or get_attr(layer, 'epsilon')
    # TODO: decide what to do with 'is_test' attribute

    if auto_pad and not auto_pad in known_paddings:
        print('IGNORED: unknown padding', auto_pad)

    if size == [None, None]:
        size = None
    if size: size = np.array(size).astype(int).tolist()

    o_l.activation  = known_activations.get(activation) or 0
    o_l.pads        = known_paddings.get(auto_pad) if auto_pad else pads or starts or [0,0,0,0]
    o_l.strides     = strides or slice_strides or []
    o_l.pool_size   = pool_size or size or shape or ends or []
    o_l.axis        = embody(axis, default=-1)
    o_l.alpha       = embody(alpha, default=1)
    o_l.beta        = beta or 0


    # Patch shapes & data
    try:
        tensor_names = [i for i in inputs if i in model_tensors]
        o_l.tensors = [
            Struct(name=model_tensors[x].name, shape=model_tensors[x].dims, data=get_tensor_data(model_tensors[x]))
            for x in tensor_names]
        o_l.inputs = [i for i in inputs if i not in model_tensors]

        shapes = klass.patch_shapes([x.shape for x in o_l.tensors])

        # if we have more shapes than actual tensors,
        # then create & fill missing tensors with zeros
        in_tensor_num = len(o_l.tensors)
        for index, new_shape in enumerate(shapes):
            if index >= in_tensor_num:
                new_tensor = Struct(name = ('/model_weights/%s/%s/patch:%i') % (name, name, index-in_tensor_num),
                    shape = new_shape,
                    data = np.zeros(new_shape))
                o_l.tensors.append(new_tensor)
        assert(len(shapes) <= len(o_l.tensors))

        if hasattr(klass, 'patch_data'):
            data = [x.data for x in o_l.tensors]

            patch_data_fn = klass.patch_data
            patch_data_expected_arg_count = patch_data_fn.__code__.co_argcount
            patch_data_args = (data, layer) if patch_data_expected_arg_count > 1 else (data,)
            tensor_data = patch_data_fn(*patch_data_args)
            o_l.tensors = o_l.tensors[:len(tensor_data)] # resize tensor array to match patched data - patching might reduce number of tensors
            for x, data in zip(o_l.tensors, tensor_data):
                x.data = data

        # after this point we should have equal amount of shapes and tensors
        assert(len(o_l.tensors) == len(shapes))

        for x, shape in zip(o_l.tensors, shapes):
            x.shape = shape

    except AttributeError:
        # no 'patch_data' lambda was specified, op does not require tensor args
        o_l.tensors = []
        o_l.inputs = inputs

    try:
        attrs = klass.patch_attrs(o_l, layer)
        for k, v in attrs.items():
            o_l.__dict__[k] = v
    except AttributeError:
        pass # no 'patch_attrs' lambda was specified

    # Force all tensors to float32
    for x in o_l.tensors:
        x.data = x.data.astype(np.float32)

    # Layer is ready
    context.layers.append(o_l)

#########################################################

def convert(source_file, target_file, trim_unused_by_output="", verbose=False, compress_f16=False):
    """
    Converts a ONNX model into a Barracuda model.
    :param source_file: The ONNX Model
    :param target_file: The name of the file the converted model will be saved to
    :param trim_unused_by_output: The regexp to match output nodes to remain in the model. All other uconnected nodes will be removed.
    :param verbose: If True, will display debug messages
    :param compress_f16: If true, the float values will be converted to f16
    :return:
    """
    if (type(verbose)==bool):
        args = Struct()
        args.verbose = verbose
        args.print_layers = verbose
        args.print_source_json = verbose
        args.print_barracuda_json = verbose
        args.print_layer_links = verbose
        args.print_patterns = verbose
        args.print_tensors = verbose
    else:
        args = verbose

    if args.print_supported_ops:
        barracuda.print_known_operations(known_classes, known_activations)

    # Load ONNX model
    print("Converting %s to %s" % (source_file, target_file))
    i_model = onnx.load(source_file)

    if args.print_source_json or args.verbose:
        for layer in i_model.graph.node:
            print(MessageToJson(layer) + ",")

    # Convert
    o_model = barracuda.Model()
    o_model.layers, o_input_shapes, o_model.tensors, o_model.memories = \
        process_model(i_model, args)

    # Trim
    if trim_unused_by_output:
        o_model.layers = barracuda.trim(o_model.layers, trim_unused_by_output, args.verbose)


    # Find model inputs & outputs
    all_inputs = {i for l in o_model.layers for i in l.inputs}

    # Create load layers for constants
    const_tensors = [i for i in all_inputs if i in o_model.tensors]
    const_tensors += o_model.globals
    for x in const_tensors:
        shape = adapt_input_shape(o_model.tensors[x].dims) if hasattr(o_model.tensors[x], 'dims') and len(o_model.tensors[x].dims) > 0 else [1, 1, 1, 1]

        o_l = Struct(
            type=255,  # Load
            class_name="Const",
            name=x,
            pads=[0, 0, 0, 0],
            strides=[],
            pool_size=[],
            axis=-1,
            alpha=1,
            beta=0,
            activation=0,
            inputs=[],
            tensors=[Struct(
                name=x,
                shape=shape,
                data=np.reshape(get_tensor_data(o_model.tensors[x]), shape).astype(np.float32))]
        )
        o_model.layers.insert(0, o_l)

    all_layers = {l.name for l in o_model.layers}

    # global inputs - are inputs that are NOT connected to any layer in the network
    # global outputs - are outputs that are NOT feeding any layer in the network
    o_model.inputs = {i:o_input_shapes[i] for l in o_model.layers for i in l.inputs if i not in all_layers}
    o_model.outputs = [l.name for l in o_model.layers if l.name not in all_inputs]

    # Compress
    if compress_f16:
        o_model = barracuda.compress(o_model)

    # Summary
    barracuda.summary(o_model,
        print_layer_links = args.print_layer_links or args.verbose,
        print_barracuda_json = args.print_barracuda_json or args.verbose,
        print_tensors = args.print_tensors or args.verbose)

    # Write to file
    barracuda.write(o_model, target_file)
    print('DONE: wrote', target_file, 'file.')
